qshell.1
.\" QSHELL MANUAL - WRITTEN BY CHRIS PONTICELLO
.TH "QSHELL" "1" "2013-08-19" "COMP3301 Assignment 1" "User Commands"
.SH NAME
qshell \- COMP3301 Operating Systems Shell project
.SH SYNOPSIS
.B qshell
[\fIFILE\fR]
.SH DESCRIPTION
\fBqshell\fR is a minimalist shell program. It works similarly to \fBbash\fR, prompting for a command line and running commands, but does not have many of the special features of the \fBbash\fR shell.
.PP
\fBqshell\fR supports redirection of standard input/output (< >), use of unnamed pipes (|), background processes (&) and comments (#). It includes two built-in commands, \fBexit\fR and \fBcd\fR.
.PP
If a \fIFILE\fR argument is given, commands are read from the \fIFILE\fR; otherwise, \fBqshell\fR launches in interactive mode and prompts for commands.
.SH OPTIONS
\-
.SH EXIT STATUS
.TP
0
if OK,
.TP
1
for minor problems (e.g. cannot access \fIFILE\fR given)
.TP
2
for serious trouble (e.g. memory issues, internal errors)
.SH ENVIRONMENT
The behaviour of \fBqshell\fR is affected by the following environment variables.
.IP "\fBPATH\fR" 
\fBqshell\fR uses the PATH variable to look for commands.
.IX Item "PATH"
.IP "\fBHOME\fR" 
When the command \fBcd\fR is used without arguments, the shell changes directories to the directory specified by HOME.
.IX Item "HOME"
.SH FILES
.IX Header "FILES"
.IP "\fBqshell\fR" 
.IX Item "qshell"
The executable binary.
.IP "No external files are required" 
.IX Item "No external files"
.SH "SUPPORTED SHELL SYNTAX"
.IX Header "SUPPORTED SHELL SYNTAX"
.IP "command [arg1 arg2 ...] [< input_file] [> output_file] [&]" 
or
.IX Item "cmdline1"
.IP "command [arg1 arg2 ...] [> output_file] [< input_file] [&]" 
or
.IX Item "cmdline2"
.IP "command [arg1 ...] [< input_file] | command2 [arg1 ...] [> output_file]" 
 
.IX Item "cmdline3"
.SH INBUILT COMMANDS
.PP
The following commands have inbuilt definitions, which override any available programs of the same name available on the PATH. 
.PP
They do not follow \fBSHELL SYNTAX\fR as above in that they do not support pipes, backgrounding or input/output redirection.
.TP
\fBcd\fR [\fIDIR\fR]
Change directory to \fIDIR\fR if supplied; else change directory to HOME.
.TP
\fBexit\fR
Exits the shell, ignoring any arguments
.SH NOTES
.PP
In comparison to \fBbash\fR, \fBqshell\fR has many limitations. Notable ones include:
.TP
\-
Chaining of pipes (|) is not supported. Only one pipe per command line
.TP
\-
Background execution (&) may not be used with piped commands
.TP
\-
Only space characters count as whitespace. Tabs and other whitespace characters are treated as text commands/arguments.
.TP
\-
The shell does not support quoting (" or '). All spaces are treated as argument separators.
.TP
\-
Command lines longer than 128 characters or with more than 20 space-separated arguments are not supported.
.PP
Pressing Ctrl-C will terminate the foreground command currently running, but will not terminate the shell or background processes.
.SH BUGS
Pipes are currently not implemented.
.SH EXAMPLES
.IX Header "EXAMPLES"
.IP "\fBInvocation\fR" 
qshell commands.sh
.IX Item "Invocation"
.IP "\fBCommands\fR" 
cd /usr/bin
.br
cd
.br
# Now in HOME directory
.br
 
.br
ls -l > files.txt
.br
cat < files.txt
.br
grep qshell > qfiles.txt < files.txt
.br
# qfiles.txt now lists filenames from files.txt containing 'qshell'
.br
 
.br
ping -c 2 8.8.8.8 > google-dns-ping.txt &
.br
# Now pinging 8.8.8.8 in the background
.br
 
.br
date | cut -d: -f1,2 > shorttime.txt
.br
# shorttime.txt now contains the time in \fIWkday Month Day HH:MM\fR format
.IX Item "Commands"
.SH AUTHOR
Written by Chris Ponticello. christopher.ponticello@uqconnect.edu.au
.SH SEE ALSO
.TP
bash(1), cd(1), exit(1)


qshell.c
//
//  qshell.c
//

#include "qshell.h"

/* ------------------------------------------------------------------------- */
/* Top level */

int main (int argc, char * argv[]) {
	#if DEBUG>1
	fprinterr(" === QSHELL-%d === \n\n", getpid());
	#endif
	
	init();
	parse_args(argc, argv);	
	return 0;
}

void init (void) {
	/*
	 * Setup: signal handling, memory allocation, global vars
	 */
	input = stdin;
	sig_setup();
	bgPids = (int *)malloc(sizeof(int)*bgPidBufsize);
	memset(bgPids, -1, sizeof(int)*bgPidBufsize);
	for (int i=0; i<2; i++) 
		fgPid[i]=-1;
}

void stop (void) {
	/*
	 * About to close shell
	 *  - stop background processes
	 *  - free memory, close files
	 *  - exit with status 0 (OK)
	 */
	int status;

	proc_reap();
	for (int i=0; i<2; i++) 
		if (fgPid[i]!=-1) {
			proc_stop(fgPid[i]);
			waitpid(fgPid[i], &status, 0);
			fgPid[i]=-1; 
		}

	free(bgPids);
	if (input!=stdin) fclose(input);

	#if DEBUG>0
	fprinterr("\nExiting, status 0\n");
	#else
	fprinterr("\n");
	#endif
	exit(0);
}

void parse_args (int argc, char * argv[]) {
	/*
	 * Decide on action to take at startup
	 */
	if (argc==1) {
		/* No parameters - start interactive input */
		input_read();	
	} else if (argc==2) {
		/* One parameter - open as file to read commands from */
		input_load(argv[1]);
		input_read();
	} else {
		/* 2+ paramters - not supported */
		printf(USAGE);
	}
}

/* ------------------------------------------------------------------------- */
/*input handling */

void input_load (char * fname) {
	/*
	 * Set the input stream to a file pointer to FNAME
	 */
	input = fopen(fname,"r");
	if (input==NULL) {
		fprintf(stderr, "Unable to open commands file '%s': %s\n",
			fname, strerror(errno));
		exit(1);
	}
}

void input_read (void) {
	/*
	 * Main event loop to collect commands from input lines
	 */
	char c, strbuf[256], *bufp;
	char *argv[20], **argp;
	int charc, argc;

	fd_set fds;
	struct timeval timeout;
	int readyStreams;

	while (1) {
		ctrlc=0;

		/* 1. Print prompt (interactive mode) */
		if (input==stdin) {
			if (getcwd(strbuf,sizeof(strbuf))==NULL) {
				fnerror("getcwd");
				exit(2);
			}
			fprintout("[%s]? ",strbuf);
		}

		/* 2. Read from input stream.
		 */
		charc=0;
		c='\0';
		// before actually reading anything, select() on input
		//   this will allow us to pick up ctrl-c (SIGINT)
		//   without getting stuck in the blocking fgetc() loop
		while (!ctrlc) {
			FD_ZERO(&fds);
			FD_SET(fileno(input), &fds);
			timeout.tv_sec = 0;
			// block for some ms at a time
			timeout.tv_usec = 10000;
			readyStreams = select(fileno(input)+1, &fds, NULL, 
				NULL, &timeout); 
			// ignore signals interrupting our system call
			if (readyStreams<0) {
				if (errno == EINTR) continue;	
				fnerror("select");
				exit(2);
			}
			// ready
			if (!readyStreams) break;
		}

		if (!ctrlc) {
			// read up to 128 chars. Split on EoF/NL/comment(#)
			//   yes, I know I could have used readline(). 
			//   I wanted to try some new things!
			while ((c=fgetc(input))!='\n' && c!='#' && \
				!feof(input) && (charc<128)) {
				strbuf[charc++]=c;
			}
			strbuf[charc]='\0';
			// handle ctrl-c
			if (ctrlc) fprintout("control-c!");
			// flush characters until the next newline
			if (charc==128 || c=='#')
				while (fgetc(input)!='\n' && !feof(input));
			#if DEBUG>0
			if (input!=stdin)
				fprintout("%s\n", strbuf);
			#endif
			// warning for too many characters
			if (charc==128) fprinterr(WARN_CMDLINE_CHARS);
	
			/* 3. Separate arguments. Max 20 arguments. */
			for (int i=0; i<20; i++)
				argv[i]=NULL;
			argc=0;
			argp=argv;
			bufp=strbuf;
			for (; (*argp=strsep(&bufp, " "))!=NULL; )
				if (**argp!='\0') {
					argp++;
					argc++;
					if (argp>&argv[20]) 
						break;
				}	
			//warning for too many arguments
			if (argc>20) {
				argc=20;
				fprinterr(WARN_CMDLINE_ARGS);
			}
	
			/* 4. Process input */
			if (argc>0) input_parse(argc, argv);
		}	
		// flush characters until the next newline
		if (charc==128 || c=='#')
			while (fgetc(input)!='\n' && !feof(input));

		/* 5. Print background process output/termination */
		proc_reap();

		/* 6. STOP before we prompt again - is it EoF?
		 *     - For interactive mode, this is any zero-char read
		 *     - For file mode, we want to stop anytime we hit feof
		 */
		if (((charc==0) && (c!='\n') && (c!='#') && !ctrlc) || \
		    (input!=stdin && feof(input)))
			stop();
	}
}

void input_parse (int argc, char * argv[]) {
	/*
	 * Handle a list of arguments from an input line
	 */
	int pipePos=-1, inDirectPos=-1, outDirectPos=-1, backPos=-1;
	int command1Pos=0, command2Pos=-1, command1End=-1, command2End=-1;
	int consecArgc=0;
	char * argPtr;

	#if DEBUG>1
	fprintf(stderr, "argc=%d\n", argc);
	for (int i=0; i<argc; i++)
		fprinterr("arg[%d]=%s\n", i, argv[i]);
	#endif
	// 1. check for inbuilt commands "cd" and "exit"
	if (streq(argv[0], "exit")) {
		stop();
	} else if (streq(argv[0], "cd")) {
		// >1 args: a path was provided (ignore extra args)
		if (argc>1 && !streq(argv[1],"~")) {
			argPtr = argv[1];			
		// 1 args: no path provided, go to HOME
		} else {
			if ((argPtr = getenv("HOME"))==NULL)
				fnerror("getenv");
		}
		// chdir
		if (chdir(argPtr)<0)
			 fnerror("chdir");
		return;
	}

	// 2. not inbuilt: loop through the arguments for syntax validation`
	else for (int i=0; i<argc; i++) {
		// pipe (|)
		if streq(argv[i], "|") {
			// enforce correct location
			if (pipePos>=0 || i==argc-1 || consecArgc==0 ||\
				outDirectPos>=0) {
				fprintout(WARN_CMDLINE_SYNTAX);
				return;
			}
			// log info about argument structure
			if (command2Pos==-1 && command1End==-1)
				command1End=i-1;
			consecArgc=0;
			command2Pos=i+1;
			pipePos=i;
		}
		// input redirect (<)
		else if streq(argv[i], "<") {
			// enforce correct location
                        if (inDirectPos>=0 || i==argc-1 || consecArgc==0 ||\
				pipePos>=0)  {
                                fprintout(WARN_CMDLINE_SYNTAX);
                                return;
                        }
			// log info about argument structure
			if (command2Pos==-1 && command1End==-1)
				command1End=i-1;
			consecArgc=0;
			consecArgc=0;
			inDirectPos=i;
		}
		// output redirect (>)
		else if streq(argv[i], ">") {
			// enforce correct location
                        if (outDirectPos>=0 || i==argc-1  || consecArgc==0)
                                fprintout(WARN_CMDLINE_SYNTAX);
                                return;
			// log info about argument structure
			if (command2Pos==-1 && command1End==-1) {
				command1End=i-1;
			} else if (command2End==-1) {
				command2End=i-1;
			}	
			consecArgc=0;
			consecArgc=0;
			outDirectPos=i;

		}
		// background (&)
		else if streq(argv[i], "&") {
			// enforce correct location
                        if (backPos>=0 || i<argc-1 || consecArgc==0) {
                                fprintout(WARN_CMDLINE_SYNTAX);
                                return;
                        }
			// log info about argument structure
			if (command2Pos==-1 && command1End==-1) {
				command1End=i-1;
			} else if (command2End==-1) {
				command2End=i-1;
			}
			consecArgc=0;
			consecArgc=0;
			backPos=i;

		}
		// other: command/arg
		else {
			consecArgc++;
			// input/output redirection takes only 1 arg
			if (consecArgc==2 && i>=2) {
				if (outDirectPos==i-2 || inDirectPos==i-2) {
					fprintout(
						WARN_CMDLINE_SYNTAX);
					return;
				}
			} 
		}
	}
	// Close off end of commands
	if (command1End==-1)
		command1End=argc-1;
	if (command2Pos>=0 && command2End==-1)
		command2End=argc-1;

	// 3 execute!
	#if DEBUG>1
	fprinterr("exec com1=%d..%d com2=%d..%d in=%d "\
		"out=%d pipe=%d\n", command1Pos, command1End, 
		command2Pos, command2End, inDirectPos, outDirectPos,
		pipePos);
	#endif

	input_exec(command1End-command1Pos+1,
		&argv[command1Pos],
		(command2Pos==-1?0:command2End-command2Pos+1),
		(command2Pos==-1?NULL:&argv[command2Pos]),
		(inDirectPos==-1?NULL:argv[inDirectPos+1]),
		(outDirectPos==-1?NULL:argv[outDirectPos+1]),
		(backPos==-1?0:1)
	);
	return;
}

void input_exec (int arg1c, char * arg1v[], int arg2c, char * arg2v[],
        char * inFname, char * outFname, bool background) {
	/*
	 * Execute parsed/validated input commands
	 */
	int i;
	int status;
	int fdPipe[2]; //idx 0: read, 1: write
	char *exec1v[21], *exec2v[21];
	pid_t pid1=-1, pid2=-1;

	#if DEBUG>0
	fprinterr("exec\n");
	#endif
	// allocate null-terminated argument vectors
	for (i=0; i<arg1c+1; i++)
		exec1v[i] = arg1v[i];
	exec1v[i-1]=NULL;
	if (arg2c>0) {
		for (i=0; i<arg2c+1; i++)
			exec2v[i] = arg2v[i];
		exec2v[i-1]=NULL;
	}
	// create pipe(s) (command1->command2)
	if (arg2c) pipe(fdPipe);

	// fork and execute command 1
	if (!(pid1 = fork())) {
		// CHILD 1
		// 0. Cancel signal handlers from qshell
		sig_cancel();
		// 1. Setup input/output files (if applicable)
		//      background processes redirect std[in|out|err]
		//      to /dev/null unless otherwise specified
		if (background && inFname==NULL) {
			proc_set_stream("/dev/null", "stdin", "r", stdin);
		} else {
			proc_set_stream(inFname, "stdin", "r", stdin);
		}
		if (background && outFname==NULL) {
			proc_set_stream("/dev/null", "stdout", "w", stdout);
		} else if (!arg2c) {
			proc_set_stream(outFname, "stdout", "w", stdout);
		}
		if (background)
			proc_set_stream("/dev/null", "stderr", "w", stderr);
		// 2. Redirect pipe write end (if applicable)
		if (arg2c) {
			dup2(fdPipe[1], fileno(stdout));
			close(fdPipe[0]);
		}
		// 3. Exec. Error code will not run if exec succeeds
		execvp(exec1v[0], &exec1v[0]);
		fnerror("exec");
		exit(EXIT_FAILURE);
	}

	// fork and execute any piped command
	if (arg2c)
		if (!(pid2 = fork())) {
			// CHILD 2
			// 0. Cancel signal handlers from qshell
			sig_cancel();
			// 1. Setup output file (if applicable)
			proc_set_stream(outFname, "stdout", "w", stdout);
			// 2. Redirect pipe read end
			dup2(fdPipe[0], fileno(stdin));
			close(fdPipe[1]);
			// 3. Exec. Error code will not run if exec succeeds
			execvp(exec2v[0], &exec2v[0]);
			fnerror("exec");
			exit(EXIT_FAILURE);
		}

	// PARENT
	// 0. Close pipes
	if (arg2c) {
		close(fdPipe[0]);
		close(fdPipe[1]);
	}
	// 1. Foreground: store PID(s) and wait/reap 
	if (!background) {
		fgPid[0]=pid1;
		fgPid[1]=pid2;
		#if DEBUG>0
		fprinterr("FG      [%d]\t%s\n", pid1, arg1v[0]);
		if (arg2c) 
			fprinterr("FG-Pipe [%d]\t%s\n", pid2, arg2v[0]);
		#endif
		waitpid(fgPid[0], &status, 0);
		fgPid[0]=-1; 
		if (arg2c) {
			waitpid(fgPid[1], &status, 0);
			fgPid[1]=-1; 
		}
	} else {
		// 2. Background: ----------
		proc_background_add(pid1);
		fprintout("BG      [%d]\tCommand : %s\n", pid1, arg1v[0]);
	}
	
	#if DEBUG>0
	fprinterr("\\exec\n");
	#endif
}


/* ------------------------------------------------------------------------- */
/* process handling */

void proc_set_stream (char * fname, char * sname, char * mode, FILE * stream) {
	/*
	 * Replace stdin, stdout, stderr with files
	 */
	if (fname) {
		//#if DEBUG>0
		//fprinterr("%s file : %s\n", sname, fname);
		//#endif
		if (!freopen(fname, mode, stream)) {
			fnerror("freopen");
			exit(EXIT_FAILURE);
		}
	}
}

void proc_reap (void) {
	/*
	 * Reap (clean up) background processes.
	 */
	int i;
	int pid, status;
	
	#if DEBUG>1
	fprinterr("\nBackground PID table:\n");
	#endif
	// Scan bgPids for running processes
	for (i=0; i<bgPidBufsize; i++) {
		pid = bgPids[i];
		if (pid!=-1) {
			#if DEBUG>1
			fprinterr("-- %d\n", pid);
			#endif
			// If nonblocking wait(reap) succeeds, go to the
			//   reaper handler & clear the background PID
			if (waitpid(pid, &status, WNOHANG)>0) {
				proc_do_reaped(pid, status);
				bgPids[i]=-1;
			}
		}
	}
}

void proc_do_reaped(int pid, int status) {
	/*
	 * Print info on a reaped process
	 */
	fprintout("BG-Done [%d]\tStatus : %d", pid, WEXITSTATUS(status));
	if (WTERMSIG(status))
		fprintout("\tSignalled : %d-%s", WTERMSIG(status),
			strsignal(WTERMSIG(status)));
	fprintout("\n");
}

void proc_stop (int pid) {
	/*
	 * Terminate process with given pid.
	 *   Ignore if PID<0
	 */
	if (pid>=0) {
		kill(pid, SIGNAL_STOP);
		#if DEBUG>0
		fprinterr("Terminating process : %d\n", pid);
		#endif
	}
}

void proc_background_add (int pid) {
	/*
	 * Add a PID to the background process buffer.
	 *   grow buffer if needed
	 */
	int i;
	int * newBgPids;

	// Try to write to an existing empty slot.
	//   PID<0 is considered empty
	for (i=0; i<bgPidBufsize; i++) {
		if (bgPids[i]==-1) {
			bgPids[i]=pid;
			break;
		}
	}	
	// Expand memory if needed - copy to a new buffer, swap/free pointers
	if (i==bgPidBufsize) {
		bgPidBufsize+=BUFFER_INCREMENT;
		newBgPids = (int *)malloc(sizeof(int)*bgPidBufsize);
		memset(newBgPids, -1, sizeof(int)*bgPidBufsize);
		memcpy(newBgPids, bgPids, 
			sizeof(int)*(bgPidBufsize-BUFFER_INCREMENT)); 
		free(bgPids);
		bgPids=newBgPids;
		// re-call function - loop again to place the pid
		proc_background_add(pid);
		return;
	}

	#if DEBUG>1
	fprinterr("<bg-%d>\n", pid);
	#endif
}


/* ------------------------------------------------------------------------- */
/* signal handling */

void sig_do_int (int status) {
	/*
	 * Interrupt on SIGINT
	 */
	#if DEBUG>0
	fprinterr("SIGINT(%d)\n", status);
	#else
	fprintout("\n");
	#endif
	ctrlc=1;

	// Kill currently foreground processes (if any)
	for (int i=0; i<2; i++)
		proc_stop(fgPid[i]);
}

void sig_do_child (int status) {
        /*
         * Debug handler so we can see when children signal completion.
	 * We do reaping elsewhere (input loop, just before prompt)
         */
	#if DEBUG>0
	fprinterr("SIGCHLD(%d)\n", status);
	#endif
}

void sig_do_shutdown (int status) {
        /*
         * Interrupt on SIGTERM - gracefully shutdown
         */
	#if DEBUG>0
	fprinterr("SIGTERM(%d)\n", status);
	#endif
	stop();
}

void sig_do_pipe (int status) {
        /*
         * Debug handler in case we get sigpipe from stream writes/reads
         */
	#if DEBUG>0
	fprinterr("SIGPIPE(%d)\n", status);
	#endif
}

void sig_setup (void) {
        /*
         * Enable interrupts for signal handling
         */
	struct sigaction sa[4];
	
	// SIGINT (ctrl-c) kills active process
	sa[0].sa_handler = sig_do_int;
	sa[0].sa_flags = SA_RESTART;
	sigaction(SIGINT, &sa[0], &storedSigActions[0]);
	sa[1].sa_handler = sig_do_child;
	sa[1].sa_flags = SA_RESTART;
	sigaction(SIGCHLD, &sa[1], &storedSigActions[1]);
	sa[2].sa_handler = sig_do_shutdown;
	sa[2].sa_flags = SA_RESTART;
	sigaction(SIGTERM, &sa[2], &storedSigActions[2]);
	sa[3].sa_handler = sig_do_pipe;
	sa[3].sa_flags = SA_RESTART;
	sigaction(SIGPIPE, &sa[3], &storedSigActions[3]);
}

void sig_cancel (void) {
        /*
         * Restore signal handlers stored in storedSigActions
	 * Used by child processes
         */
	sigaction(SIGINT,  &storedSigActions[0], 0);
	sigaction(SIGCHLD, &storedSigActions[1], 0);
	sigaction(SIGTERM, &storedSigActions[2], 0);
	sigaction(SIGPIPE, &storedSigActions[3], 0);
}

void sig_block (int signal) {
        /*
         * Block handling for a signal type
         */
	sigemptyset(&sigList);
	sigaddset(&sigList, signal);
	sigprocmask(SIG_BLOCK, &sigList, NULL);
}

void sig_unblock (void) {
        /*
         * Unblock handling for a signal type
         */
	sigprocmask(SIG_UNBLOCK, &sigList, NULL);
	sigemptyset(&sigList);
}

/* ------------------------------------------------------------------------- */

